server/src/main/scala/org/http4s/server/websocket/package.scala

```diff
@@ -37,12 +38,23 @@ package object websocket {
     *                 the case of a normal disconnection such as a Close handshake or due to a connection error. There
     *                 are plans to address this limitation in the future.
     * @param status The status code to return to a client making a non-websocket HTTP request to this route
+    * @param headers Websocket Handshake response headers, such as Sec-WebSocket-Protocol
     */
+  def WS[F[_]](
```

Comment:
`WS` is kind of a strange name.  I think we could deprecate the old `WS` constructors and make the one we really want here.  I still like:

```
def WebSocketResponse(
    send: Stream[F, WebSocketFrame],
    receive: Sink[F, WebSocketFrame],
    headers: Headers = Headers.empty, // Sec-WebSocket-Protocol, Cookies, etc. go here
    onNonWebSocketRequest: F[Response[F]] = DefaultNonWebSocketResponse,
    onHandshakeFailure: F[Response[F]] = DefaultHandshakeFailureResponse)
```

That arranges the default arguments in the order they're most likely to be overridden.  If you want to set aside the custom handshake failure for now, you can just leave that off the end (as well as the context, which you've already done).

Reply:
I need your help with the proper definition of DefaultNonWebSocketResponse, tried to do it with def so I can specify the F as its defined on def WS, but I think I got the implicit wrong.

Reply:
I think I should be doing
```
private def DefaultNonWebSocketResponse[F[_]](implicit F: Monad[F]) = Response[F](Status.NotImplemented).withBody("This is a WebSocket route.")
```

Reply:
That looks right.  You could alternately use the shorthand:

`private def DefaultNonWebSocketResponse[F: Monad] = `

Your syntax gives you access to `F` if you want to invoke methods directly on the type class.  The shorthand just guarantees that `F` has a `Monad`.  I usually use the shorthand unless I need the named version in the definition.

Reply:
my issue then is with the using it as default param value:

```
onNonWebSocketRequest: F[Response[F]] = DefaultNonWebSocketResponse
```

getting 
```
Error:(59, 66) ambiguous implicit values:
 both value catsInstancesForId in package cats of type => cats.Bimonad[cats.package.Id] with cats.CommutativeMonad[cats.package.Id] with cats.Comonad[cats.package.Id] with cats.NonEmptyTraverse[cats.package.Id]
 and method catsStdInstancesForEither in trait EitherInstances of type [A]=> cats.MonadError[[β$0$]scala.util.Either[A,β$0$],A] with cats.Traverse[[β$1$]scala.util.Either[A,β$1$]]
 match expected type cats.Monad[F]
                         onNonWebSocketRequest: F[Response[F]] = DefaultNonWebSocketResponse,
```

Reply:
this is the entire method signature 

```
  def WebSocketResponseWS[F[_]](
                         send: Stream[F, WebSocketFrame],
                         receive: Sink[F, WebSocketFrame],
                         headers: Headers = Headers.empty, // Sec-WebSocket-Protocol, Cookies, etc. go here
                         onNonWebSocketRequest: F[Response[F]] = DefaultNonWebSocketResponse,
                         onHandshakeFailure: F[Response[F]] = DefaultHandshakeFailureResponse)(implicit F: Functor[F]): F[Response[F]]
```

Reply:
Oh.  Hmm.  This is tricky.  `onNonWebsocketRequest` needs a `Monad[F]` instance.  You can provide one to `WebSocketResponseWS` (you currently have a `Functor[F]`), but that implicit doesn't apply to a defaul argument.

I'm going to sleep, but I started [rambling on Gitter about here](https://gitter.im/http4s/http4s?at=5a3b395f540c78242ded39a3).

Reply:
The way this is handled other places is to have the constructor not take default arguments and then having and `apply` method in the companion object that has default parameters, which allows for type class constraints on the defaults.

Reply:
made it work with the builder approach and then saw @aeons comment
I am not getting it right with the companion object approach...
```
import cats._
import cats.implicits._
import fs2._
import org.http4s.websocket.WebsocketBits.WebSocketFrame
import org.http4s.websocket.{WebSocketContext, Websocket}
import org.http4s.{AttributeEntry, Headers, Response, Status}

case class WebSocketBuilder[F[_]](
                                   send: Stream[F, WebSocketFrame],
                                   receive: Sink[F, WebSocketFrame],
                                   headers: Headers,
                                   onNonWebSocketRequest: F[Response[F]],
                                   onHandshakeFailure: F[Response[F]]) {

  def build(send: Stream[F, WebSocketFrame],
            receive: Sink[F, WebSocketFrame],
            onNonWebSocketRequest: F[Response[F]],
            onHandshakeFailure: F[Response[F]])(implicit F: Monad[F]): F[Response[F]] = {
    onNonWebSocketRequest.map(
      _.withAttribute(AttributeEntry(
        websocketKey[F],
        WebSocketContext(Websocket(send, receive), headers, onHandshakeFailure))))
  }
}

object WebSocketBuilder {

  def apply[F[_]](send: Stream[F, WebSocketFrame],
                  receive: Sink[F, WebSocketFrame],
                  headers: Headers = Headers.empty,
                  onNonWebSocketRequest: F[Response[F]] = Response[F](Status.NotImplemented).withBody("This is a WebSocket route."),
                  onHandshakeFailure: F[Response[F]] = Response[F](Status.BadRequest).withBody("WebSocket handshake failed."))
                 (implicit F: Monad[F]): F[Response[F]] =
    WebSocketBuilder(send, receive, headers, onNonWebSocketRequest, onHandshakeFailure)
}
```


```
Error:(33, 102) could not find implicit value for parameter F: cats.Monad[F]
                  onNonWebSocketRequest: F[Response[F]] = Response[F](Status.NotImplemented).withBody("This is a WebSocket route."),
```

Reply:
This is a minimisation of the problem, let's assume:
```scala
  import cats._, implicits._
  def default[F[_]: Monad]: F[Int] = 1.pure[F]
```
then:
```scala
  case class No[F[_]: Monad](s: String, f: F[Int] = default[F])
```
Does not compile. You can do:
```scala
  case class Yes[F[_]](s: String, f: F[Int])
  object Yes {
    class Builder[F[_]: Monad] {
      def apply(s: String, f: F[Int] = default[F]) = Yes(s, f)
    }

    def apply[F[_]: Monad] = new Builder[F]
  }

  def yes = Yes[Option].apply("yes") // Yes(yes, Some(1))
```
Ideally, we should be able to call `yes` with Yes[Option]("yes"),
but when we do that the compiler thinks we are passing evidence for
Monad explicitly, and complains. This is annoying, and there's a
proposal to fix it in the language (the `explicitly` keyword)

There is a solution if you're willing to crank up the magic though...
```scala
  // you lose apply, which you want to lose since it's overriden
  // you lose copy, which in this case it kinds sucks
  // you still have equals, hash and toString
  sealed abstract case class Magic[F[_]](s: String, f: F[Int])
  object Magic {
    // same builder as before
    class Builder[F[_]: Monad] {
      def apply(s: String, f: F[Int] = default[F]) = new Magic(s, f){}
    }
    // magic
    class Carrier[F[_]]
    object Carrier {
      implicit def conv[F[_]](c: Carrier[F])(implicit F: Monad[F]): Builder[F] =
        new Builder[F]
    }

    def apply[F[_]] = new Carrier[F]
  }

  def magic = Magic[Option]("magic") // Magic(magic, Some(1))
```

Reply:
Yeah, I was a bit unclear in my explanation, You can do what @SystemFw suggests or see how it's done for BlazeBuilder here: https://github.com/http4s/http4s/blob/master/blaze-server/src/main/scala/org/http4s/server/blaze/BlazeBuilder.scala#L294

So the apply on the companion object only takes the implicit evidence, and then you use the `.with*` methods on the builder to change the things you want to change from the defaults.

Reply:
fwiw, I think the best tradeoff between convenience and clarity is method no.2, except you rename `apply` to `build`, and use default parameters rather than multiple `with*` methods.
Method no.3 is interesting enough to be worth experimenting with imho, but it's a bit of a tradeoff (error messages, weird encoding)

Reply:
I tried using the partially applied approach on BlazeBuilder, but it fails if you don't set any of the arguments.

`BlazeBuilder.build[IO]` fails, `BlazeBuilder.build[IO]()` fails since it expects the `Effect` evidence there.

With the carrier, you still have to give it empty parens to make it work, which imo is worse ergonomics than what we have currently, which works with `BlazeBuilder[IO].with...`

Reply:
yeah you need `BlazeBuilder[IO].build()`. The parenthesis are unavoidable due to the way default arguments work in Scala (for some mysterious reason), it probably works best in cases where there's at least one argument that does not have a default